### ISSUE TITLE
transformation of frames doesn't always work for frames with multiple obstimes

### ISSUE THREAD

**Issue by StuartLittlefair** [astropy/astropy#4289]
The code below works exactly as I'd expect:

```
from astropy import time, coordinates as coord
now = time.Time.now()

WHT  = coord.EarthLocation(lat=28.76063889,lon=-17.88163889,height=2332.)
itrs = coord.itrs(WHT.itrs,obstime=now)
gcrs = itrs.transform_to(coord.GCRS(obstime=now))
```

but the code below doesn't work at all
```
times = time.Time([57329.906211306305,57329.92],format='mjd',scale='utc')
itrs = coord.itrs(WHT.itrs,obstime=times)
gcrs = itrs.transform_to(coord.GCRS(obstime=times))
```

In general I guess this transform could be ambiguous, but where the two times arrays are identical, or the same size this ought to be able to do something sensible I think. Fixing this would make implementation of #2244 much easier.

**Comment by mhvk** [comment]
@StuartLittlefair - your examples fails for me even for the scalar case, both in master and `1.0.5` -- at least, I'm assuming you meant `itrs = coord.ITRS(WHT, obstime=now)`.  But in master, one can do
```
itrs = WHT.get_itrs(obstime=now)
```
The latter line also works with multiple times, but then the final line gives a broadcasting error. It is due to an incorrect transpose in L55 in utils.py (in `cartrepr_from_matmul`, see below), which doesn't take into account that `pmat` can have more than 2 dimensions. There is a version below that is better, but still incorrect, in that it would break for the case of times that had more than 1 dimension (the `xyz` stanza also assumes 1-dimensional coordinates).
```
def cartrepr_from_matmul(pmat, coo, transpose=False):
    """
    Note that pmat should be an ndarray, *not* a matrix.
    """
    if pmat.shape[-2:] != (3, 3):
        raise ValueError("tried to do matrix multiplication with an array that "
                         "doesn't end in 3x3")
    if coo.isscalar:
        # a simpler path for scalar coordinates
        if transpose:
            pmat = pmat.T
        newxyz = np.sum(pmat * coo.cartesian.xyz, axis=-1)
    else:
        xyz = coo.cartesian.xyz.T
        # these expression are the same as iterating over the first dimension of
        # pmat and xyz and doing matrix multiplication on each in turn.  resulting
        # dimension is <coo shape> x 3
        pmat = pmat.reshape(pmat.size//9, 3, 3)
        if transpose:
            pmat = pmat.transpose(0, 2, 1)
        newxyz = np.sum(pmat * xyz.reshape(xyz.size//3, 1, 3), axis=-1).T

    return CartesianRepresentation(newxyz)
```

**Comment by StuartLittlefair** [comment]
Then I think this is as simple as replacing ```if coo.isscalar``` with ```if coo.isscalar and coo.obstime.isscalar``` isn't it? At least for 1D times and coordinates...

**Comment by StuartLittlefair** [comment]
@mhvk Actually, I meant to write ```itrs = coord.ITRS(WHT.cartesian, obstime=now)```, but ```itrs = WHT.get_itrs(obstime=now)``` is much more readable.

Changing ```cartrepr_from_matmul``` to work for non-scalar times is the easy bit here. The difficult bit is that there's ambiguity in how the current coordinate framework is supposed to work with multiple obstimes. Consider the GCRS frame below:

```
<GCRS Coordinate (obstime=[ 57329.90621131  57329.92      ], obsgeoloc=[ 0.  0.  0.] m, obsgeovel=[ 0.  0.  0.] m / s): (ra, dec, distance) in (deg, deg, m)
    [(350.98930115, 28.51140639, 6375550.15886377),
     (355.96653484, 28.51207325, 6375553.74963513)]>
```

I can't tell if there's a 1-1 pairing of obstimes and coordinates, or if each coordinate should have a pair of obstimes.

My own opinion is that for coordinates and obstimes to make sense, each coordinate object should only be allowed a single obstime, or (preferably) each entry in the coordinate array as an associated obstime array. I don't know what @eteq thinks of this.

**Comment by mhvk** [comment]
Given that we run numpy underneath, I think we have no choice but to have obstimes and coordinates be broadcast together to make a final set of "space-time" coordinates. So, in your example, two coordinates at different times. If one wanted the two times for each coordinate, it would be `obstime= [[time1], [time2]]`.

**Comment by mhvk** [comment]
I put the fix for `cartrepr_from_matmul` in #4123, as it is needed to allow more than 1-D coordinates anyway.  Specifically, see https://github.com/astropy/astropy/commit/efde8a565d0ea36cd18ffa099e596704e600c150

**Comment by embray** [comment]
So is this a bug?

**Comment by StuartLittlefair** [comment]
Yes, it is a bug, but I regard @mhvk as having fixed it in #4123. I guess if people agree we should close this issue.
